pkg/logging/fluentd_logger.go

```diff
@@ -0,0 +1,254 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package logging
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"math"
+	"net/url"
+	"runtime"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/containerd/containerd/runtime/v2/logging"
+	"github.com/fluent/fluent-logger-golang/fluent"
+	"github.com/pkg/errors"
+)
+
+type FluentdLogger struct {
+	Opts map[string]string
+}
+
+const (
+	fluentAddress                 = "fluentd-address"
+	fluentdAsync                  = "fluentd-async"
+	fluentdBufferLimit            = "fluentd-buffer-limit"
+	fluentdRetryWait              = "fluentd-retry-wait"
+	fluentdMaxRetries             = "fluentd-max-retries"
+	fluentdSubSecondPrecision     = "fluentd-sub-second-precision"
+	fluentdAsyncReconnectInterval = "fluentd-async-reconnect-interval"
+	fluentRequestAck              = "fluentd-request-ack"
+)
+
+const (
+	defaultBufferLimit = 1024 * 1024
+	defaultHost        = "127.0.0.1"
+	defaultPort        = 24224
+	defaultProtocol    = "tcp"
+
+	defaultMaxRetries = math.MaxInt32
+	defaultRetryWait  = 1000
+
+	minReconnectInterval = 100 * time.Millisecond
+	maxReconnectInterval = 10 * time.Second
+)
+
+type fluentdLocation struct {
+	protocol string
+	host     string
+	port     int
+	path     string
+}
+
+func (f *FluentdLogger) Process(_ string, config *logging.Config) error {
+	if runtime.GOOS == "windows" {
+		// TODO: support fluentd on windows
+		return fmt.Errorf("logging to fluentd is not supported on windows")
+	}
+	fluentConfig, err := parseFluentdConfig(f.Opts)
+	if err != nil {
+		return err
+	}
+	fluentClient, err := fluent.New(fluentConfig)
+	if err != nil {
+		return errors.Wrap(err, "create fluent client failed")
+	}
+	var wg sync.WaitGroup
+	wg.Add(2)
+	fun := func(wg *sync.WaitGroup, reader io.Reader, config *logging.Config, source string) {
+		defer wg.Done()
+		metaData := map[string]string{
+			"container_id": config.ID,
+			"namespace":    config.Namespace,
+			"source":       source,
+		}
+		scanner := bufio.NewScanner(reader)
+		for scanner.Scan() {
+			if scanner.Err() != nil {
+				return
+			}
+			metaData["log"] = scanner.Text()
+			fluentClient.PostWithTime(f.Opts[Tag], time.Now(), metaData)
+		}
+	}
+	go fun(&wg, config.Stdout, config, "stdout")
```

Comment:
`go fun(&wg, config, "stdout")` : `config` contains the Readers no need to pass them as parameters